src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/plugins/BitBlt.java

```diff
@@ -2542,10 +2542,14 @@ public Object primitiveCopyBits(final PointersObject bbObj, final long factor) {
         if (!loadBitBltFromwarping(bbObj, false)) {
             PrimitiveFailed.andTransferToInterpreter();
         }
-        copyBits(factor);
-        assert !failed();
-        showDisplayBits();
-        assert !failed();
+        try {
+            copyBits(factor);
+            assert !failed();
+            showDisplayBits();
+            assert !failed();
```

Comment:
Interesting find! Which tests caused the assertions errors? Would it be enough to turn each of these two `assert`s into this?
```java
if (failed()) {
  PrimitiveFailed.andTransferToInterpreter();
}

Reply:
but there are other asserts too, embedded in other methods, this is just the top one. Actually the asserts hit and raised were the ones that I ended up removing (in copyLoop())

Reply:
If we want to be correct, we'd need to do the `PrimitiveFailed.andTransferToInterpreter();` dance for every assertion. I've decided to use assertions to be able to debug them, because sometimes (e.g. during startup) BitBlt is supposed to fail. Let's leave all assertions in place unless they are the reasons some tests are failing or some images are crashing. Then, we should debug them or let the primitive fail.

Reply:
but doesn't my solution achieve that as well? The tests should fail as well with a primitive failure, and if you run under debug with trapping AssertionError, you catch it when it's thrown. An assert is equivalent with a vm crash, i don't think it's better than a primitive failure - this errors do happen on the vm side, inside the primitive implementation. That's what primitive failures are for

Reply:
Right, I forgot that BitBlt and Balloon are behind Truffle boundaries and therefore excluded from Truffle compilation. Shall we log the assertion error to stdout, so that users will see and can report them?